paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java

```diff
@@ -179,4 +187,32 @@ public TableCommitImpl newCommit(String commitUser) {
                 options().writeOnly() ? null : store().newExpire(),
                 options().writeOnly() ? null : store().newPartitionExpire(commitUser));
     }
+
+    private Optional<TableSchema> tryTimeTravel(Options options) {
+        CoreOptions coreOptions = new CoreOptions(options);
+
+        if (coreOptions.startupMode() == CoreOptions.StartupMode.FROM_SNAPSHOT) {
+            long snapshotId = coreOptions.scanSnapshotId();
+            if (snapshotManager().snapshotExists(snapshotId)) {
+                long schemaId = snapshotManager().snapshot(snapshotId).schemaId();
+                LOG.info("Time traveling to snapshot {}, schemaId is {}.", snapshotId, schemaId);
+                return Optional.of(schemaManager().schema(schemaId).copy(options.toMap()));
+            }
+        }
+
+        if (coreOptions.startupMode() == CoreOptions.StartupMode.FROM_TIMESTAMP) {
+            long timestamp = coreOptions.scanTimestampMills();
+            Long snapshotId = snapshotManager().earlierOrEqualTimeMills(timestamp);
```

Comment:
maybe we can just new a `StaticFromTimestampStartingScanner` to get snapshot.
Provide a `StaticFromTimestampStartingScanner.getSnapshot(SnapshotManager, long timestamp)`?